### PR TITLE
ensure that user and group IDs in LDAP's tables are also max 64chars

### LINKED ISSUE #28653: LDAP: when a value for a user/group id is too long, it should be hashed
We have a limit for user and group ids of 64 characters. At the moment, if a value from LDAP that is supposed to be used as such, is longer than this limit, there is no handling. The possibly best thing to do is to hash the id in this case and document this behaviour. As of now it would fail to write to the DB.

### PR DESCRIPTION
**Pull request by blizzz**
- limitation by core tables (e.g. sharing), IDs are always 64chars
- when longer group IDs were requested they are hashed (does not affect displaynames)

fixes #28653 

While the LDAP tables were always this long, these never seemed to have happened after a first report. So it's an edge case, the migration steps should run fairly quickly in any cases. On my test instance with >34k entries in the user mappings table (but only few long ones to check – pretty much as in reality) it finished almost immediately (pgsql).

### CODE REVIEW

File: apps/user_ldap/lib/Access.php [hunk 1]
```diff
@@ -1431,13 +1438,29 @@ public function sanitizeUsername($name) {
 		// Every remaining disallowed characters will be removed
 		$name = preg_replace('/[^a-zA-Z0-9_.@-]/u', '', $name);
 
+		if(strlen($name) > 64) {
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Access.php:1441]:
```suggestion
		if (strlen($name) > 64) {
```

File: apps/user_ldap/lib/Access.php [hunk 2]
```diff
@@ -1431,13 +1438,29 @@ public function sanitizeUsername($name) {
...
 		if ($name === '') {
 			throw new \InvalidArgumentException('provided name template for username does not contain any allowed characters');
 		}
 
 		return $name;
 	}
 
+	public function sanitizeGroupIDCandidate(string $candidate): string {
+		$candidate = trim($candidate);
+		if(strlen($candidate) > 64) {
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Access.php:1454]:
```suggestion
		if (strlen($candidate) > 64) {
```

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+		return $changeSchema ? $schema : null;
+	}
+
+	protected function handleIDs(string $table, bool $emitHooks) {
+		$q = $this->getSelectQuery($table);
+		$u = $this->getUpdateQuery($table);
+
+		$r = $q->executeQuery();
+		while($row = $r->fetch()) {
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:82]:
```suggestion
		while ($row = $r->fetch()) {
```

**Review by nickvergessen — COMMENTED** [review]
CS checker goes mimimi

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$changeSchema = false;
+		foreach (['ldap_user_mapping', 'ldap_group_mapping'] as $tableName) {
+			if ($schema->hasTable($tableName)) {
```
Comment (ChristophWurst) [inline apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:64]: as per my usual remark I don't think you need to check for existence, do you? if that table doesn't exist here then the migration system is broken. from the app's perspective you can expect previous migrations to have been run

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\DB\Types;
+use OCP\IDBConnection;
+use OCP\IUserManager;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Auto-generated migration step: Please modify to your needs!
```
Comment (ChristophWurst) [inline apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:20]: :fire:

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php (hunk 4 above)
Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:64]: that's very defensive and superfluous, you're right. I pointed out similar cases elsewhere as well.

**Review by ChristophWurst — APPROVED** [review]
Looks good

**Comment by blizzz** [comment]
Psalm should also be happy now

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+			}
+		}
+		$r->closeCursor();
+	}
+
+	protected function getSelectQuery(string $table): IQueryBuilder {
+		$q = $this->dbc->getQueryBuilder();
+		$q->select('owncloud_name', 'directory_uuid')
+			->from($table)
+			->where($q->expr()->like('owncloud_name', $q->createNamedParameter(str_pad('', 65, '_') . '%'), Types::STRING));
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:110]: May I introduce to you: https://www.php.net/manual/en/function.str-repeat.php

**Review by nickvergessen — APPROVED** [review]
🐘

File: apps/user_ldap/lib/Migration/Version1120Date20210917155206.php (hunk 6 above)
Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1120Date20210917155206.php:110]: oh, yeah, this i wanted to use and was irritated by the arguments of str_pad :grin:

**Comment by PVince81** [comment]
stable20: https://github.com/nextcloud/server/pull/28971

**Pull request merged**